In [1]:
import trainer.glue_base as glue_base
import models.sparse_token as sparse
import pickle, importlib
importlib.reload(glue_base)
importlib.reload(sparse)
Glue = glue_base.GlueAttentionApproxTrainer
PICKLE_PATH = "glue_benchmark_wiki.pkl"
TEX_PATH = "saves_plot/glue_benchmark_wiki.tex"

/home/ainl/anaconda3/envs/torch/lib/python3.8/site-packages/tqdm/auto.py:22: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [2]:
subsets = ["cola","mnli","mrpc","qnli","qqp","rte","sst2","stsb","wnli",]
kss = [
    0.1, 0.25, 0.375, 0.5, 0.625, 0.75, 0.875, 0.999, 'dynamic', 
    'dynamic:avg:avg:true', 'dynamic:avg:avg:false', 'dynamic:avg:max:true', 'dynamic:avg:max:false',
    'dynamic:max:avg:true', 'dynamic:max:avg:false', 'dynamic:max:max:true', 'dynamic:max:max:false',
]
kss = ['dynamic']
sparse.benchmark_reset()
# subsets = ["mrpc"]
# kss = ['dynamic:avg:avg:f',0.1]

def get_score(score):
    if 'accuracy' in score:
        return score['accuracy'], "acc"
    first_metric = list(score.keys())[0]
    return score[first_metric], first_metric

results = {}
i = 0
for subset in subsets:
    trainer = Glue(dataset=subset, factor=16, batch_size=-1, device=0, wiki_train=True)
    trainer.load()
    scores = {}
    metric_name = ""
    bert_score, metric_name = get_score(trainer.eval_base_model())
    scores['bert'] = f'{bert_score:.5f}'
    for ks in kss:
        sparse.benchmark_reset()
        sparse_score, _ = get_score(trainer.eval_sparse_model(ks=ks))
        if isinstance(ks, str) and ks.startswith('dynamic'):
            est_k = sparse.benchmark_get_average('est_k')
            scores[str(ks)] = f'{sparse_score:.5f} (k:{est_k:.2f})'
        else:
            scores[str(ks)] = f'{sparse_score:.5f}'
        i += 1
        count = len(subsets) * len(kss)
        print(f'{i}/{count}')
    results[f"{subset} ({metric_name})"] = scores

with open(PICKLE_PATH, 'wb') as f:
    pickle.dump(results, f)

sparse.benchmark_report()

Trainer: cola


Reusing dataset glue (/home/ainl/.cache/huggingface/datasets/glue/cola/1.0.0/dacbe3125aa31d7f70367a07a8a9e72a5a0bfeb5fc42e75c9db75b96da6053ad)
Loading cached processed dataset at /home/ainl/.cache/huggingface/datasets/glue/cola/1.0.0/dacbe3125aa31d7f70367a07a8a9e72a5a0bfeb5fc42e75c9db75b96da6053ad/cache-f0e04bff3297c11c.arrow
Loading cached processed dataset at /home/ainl/.cache/huggingface/datasets/glue/cola/1.0.0/dacbe3125aa31d7f70367a07a8a9e72a5a0bfeb5fc42e75c9db75b96da6053ad/cache-b9e7d91153cc59cf.arrow
Reusing dataset glue (/home/ainl/.cache/huggingface/datasets/glue/cola/1.0.0/dacbe3125aa31d7f70367a07a8a9e72a5a0bfeb5fc42e75c9db75b96da6053ad)
Loading cached processed dataset at /home/ainl/.cache/huggingface/datasets/glue/cola/1.0.0/dacbe3125aa31d7f70367a07a8a9e72a5a0bfeb5fc42e75c9db75b96da6053ad/cache-7e570c95fb4f22e4.arrow
Loading cached processed dataset at /home/ainl/.cache/huggingface/datasets/glue/cola/1.0.0/dacbe3125aa31d7f70367a07a8a9e72a5a0bfeb5fc42e75c9db75b96da6053ad/cac

WikitextBatchLoader: worker_main
WikitextBatchLoader: worker_main
WikitextBatchLoader: worker_main
WikitextBatchLoader: worker_main
Trainer: Save checkpoint path saves/glue-cola-16-wiki-b5.pth


eval: 100%|██████████| 131/131 [00:01<00:00, 76.11it/s] 


metric score {'matthews_correlation': 0.5338774230813111}
avg occupy 0.34522852946633237


eval: 100%|██████████| 131/131 [00:03<00:00, 42.13it/s]


metric score {'matthews_correlation': 0.04156136222110997}
avg occupy 0.34522852946633237
1/153


eval: 100%|██████████| 131/131 [00:03<00:00, 42.17it/s]


metric score {'matthews_correlation': 0.25772218156871834}
avg occupy 0.34522852946633237
2/153


eval: 100%|██████████| 131/131 [00:03<00:00, 42.17it/s]


metric score {'matthews_correlation': 0.4197091815697127}
avg occupy 0.34522852946633237
3/153


eval: 100%|██████████| 131/131 [00:03<00:00, 42.00it/s]


metric score {'matthews_correlation': 0.48571743296415204}
avg occupy 0.34522852946633237
4/153


eval: 100%|██████████| 131/131 [00:03<00:00, 41.90it/s]


metric score {'matthews_correlation': 0.5162458987125488}
avg occupy 0.34522852946633237
5/153


eval: 100%|██████████| 131/131 [00:03<00:00, 42.01it/s]


metric score {'matthews_correlation': 0.5155338105766099}
avg occupy 0.34522852946633237
6/153


eval: 100%|██████████| 131/131 [00:03<00:00, 42.30it/s]


metric score {'matthews_correlation': 0.5258867890805083}
avg occupy 0.34522852946633237
7/153


eval: 100%|██████████| 131/131 [00:03<00:00, 42.17it/s]


metric score {'matthews_correlation': 0.5338774230813111}
avg occupy 0.34522852946633237
8/153


eval: 100%|██████████| 131/131 [00:03<00:00, 39.46it/s]


metric score {'matthews_correlation': 0.5311983410233877}
avg occupy 0.34522852946633237
9/153


eval: 100%|██████████| 131/131 [00:03<00:00, 40.25it/s]


metric score {'matthews_correlation': 0.5311983410233877}
avg occupy 0.34522852946633237
10/153


eval: 100%|██████████| 131/131 [00:03<00:00, 41.04it/s]


metric score {'matthews_correlation': 0.5311983410233877}
avg occupy 0.34522852946633237
11/153


eval: 100%|██████████| 131/131 [00:03<00:00, 40.27it/s]


metric score {'matthews_correlation': 0.5311983410233877}
avg occupy 0.34522852946633237
12/153


eval: 100%|██████████| 131/131 [00:03<00:00, 40.99it/s]


metric score {'matthews_correlation': 0.5311983410233877}
avg occupy 0.34522852946633237
13/153


eval: 100%|██████████| 131/131 [00:03<00:00, 40.22it/s]


metric score {'matthews_correlation': 0.5311983410233877}
avg occupy 0.34522852946633237
14/153


eval: 100%|██████████| 131/131 [00:03<00:00, 40.49it/s]


metric score {'matthews_correlation': 0.5311983410233877}
avg occupy 0.34522852946633237
15/153


eval: 100%|██████████| 131/131 [00:03<00:00, 39.86it/s]


metric score {'matthews_correlation': 0.5311983410233877}
avg occupy 0.34522852946633237
16/153


eval: 100%|██████████| 131/131 [00:03<00:00, 40.71it/s]


metric score {'matthews_correlation': 0.5311983410233877}
avg occupy 0.34522852946633237
17/153
Trainer: mnli


WikitextBatchLoader: worker_main
WikitextBatchLoader: worker_main

WikitextBatchLoader: worker_mainWikitextBatchLoader: worker_main
Trainer: Save checkpoint path saves/glue-mnli-16-wiki-b5.pth


eval: 100%|██████████| 1227/1227 [00:26<00:00, 45.75it/s]


metric score {'accuracy': 0.8419765664798777}
avg occupy 0.2057366697815377


eval: 100%|██████████| 1227/1227 [00:30<00:00, 40.77it/s]


metric score {'accuracy': 0.48099847172694854}
avg occupy 0.2057366697815377
18/153


eval: 100%|██████████| 1227/1227 [00:31<00:00, 38.54it/s]


metric score {'accuracy': 0.7674987264391238}
avg occupy 0.2057366697815377
19/153


eval: 100%|██████████| 1227/1227 [00:34<00:00, 35.84it/s]


metric score {'accuracy': 0.8263881813550688}
avg occupy 0.2057366697815377
20/153


eval: 100%|██████████| 1227/1227 [00:35<00:00, 34.09it/s]


metric score {'accuracy': 0.8371879775853286}
avg occupy 0.2057366697815377
21/153


eval: 100%|██████████| 1227/1227 [00:37<00:00, 32.50it/s]


metric score {'accuracy': 0.8407539480387163}
avg occupy 0.2057366697815377
22/153


eval: 100%|██████████| 1227/1227 [00:39<00:00, 31.26it/s]


metric score {'accuracy': 0.8415690269994905}
avg occupy 0.2057366697815377
23/153


eval: 100%|██████████| 1227/1227 [00:41<00:00, 29.25it/s]


metric score {'accuracy': 0.8417727967396842}
avg occupy 0.2057366697815377
24/153


eval: 100%|██████████| 1227/1227 [00:43<00:00, 28.07it/s]


metric score {'accuracy': 0.8419765664798777}
avg occupy 0.2057366697815377
25/153


eval: 100%|██████████| 1227/1227 [00:34<00:00, 35.23it/s]


metric score {'accuracy': 0.8419765664798777}
avg occupy 0.2057366697815377
26/153


eval: 100%|██████████| 1227/1227 [00:34<00:00, 35.30it/s]


metric score {'accuracy': 0.8419765664798777}
avg occupy 0.2057366697815377
27/153


eval: 100%|██████████| 1227/1227 [00:34<00:00, 36.05it/s]


metric score {'accuracy': 0.8419765664798777}
avg occupy 0.2057366697815377
28/153


eval: 100%|██████████| 1227/1227 [00:34<00:00, 35.82it/s]


metric score {'accuracy': 0.8419765664798777}
avg occupy 0.2057366697815377
29/153


eval: 100%|██████████| 1227/1227 [00:33<00:00, 36.43it/s]


metric score {'accuracy': 0.8419765664798777}
avg occupy 0.2057366697815377
30/153


eval: 100%|██████████| 1227/1227 [00:34<00:00, 35.79it/s]


metric score {'accuracy': 0.8419765664798777}
avg occupy 0.2057366697815377
31/153


eval: 100%|██████████| 1227/1227 [00:33<00:00, 36.40it/s]


metric score {'accuracy': 0.8419765664798777}
avg occupy 0.2057366697815377
32/153


eval: 100%|██████████| 1227/1227 [00:34<00:00, 35.57it/s]


metric score {'accuracy': 0.8419765664798777}
avg occupy 0.2057366697815377
33/153


eval: 100%|██████████| 1227/1227 [00:34<00:00, 35.85it/s]


metric score {'accuracy': 0.8419765664798777}
avg occupy 0.2057366697815377
34/153
Trainer: mrpc


WikitextBatchLoader: worker_main
WikitextBatchLoader: worker_main

WikitextBatchLoader: worker_mainWikitextBatchLoader: worker_main
Trainer: Save checkpoint path saves/glue-mrpc-16-wiki-b5.pth


eval: 100%|██████████| 216/216 [00:02<00:00, 77.04it/s]


metric score {'accuracy': 0.8440579710144928, 'f1': 0.8865457612821593}
avg occupy 0.5399536083129833


eval: 100%|██████████| 216/216 [00:05<00:00, 42.08it/s]


metric score {'accuracy': 0.33507246376811595, 'f1': 0.0}
avg occupy 0.5399536083129833
35/153


eval: 100%|██████████| 216/216 [00:05<00:00, 42.98it/s]


metric score {'accuracy': 0.3553623188405797, 'f1': 0.08702791461412152}
avg occupy 0.5399536083129833
36/153


eval: 100%|██████████| 216/216 [00:04<00:00, 43.39it/s]


metric score {'accuracy': 0.4765217391304348, 'f1': 0.403960396039604}
avg occupy 0.5399536083129833
37/153


eval: 100%|██████████| 216/216 [00:04<00:00, 43.71it/s]


metric score {'accuracy': 0.664927536231884, 'f1': 0.7068965517241379}
avg occupy 0.5399536083129833
38/153


eval: 100%|██████████| 216/216 [00:04<00:00, 43.64it/s]


metric score {'accuracy': 0.8005797101449276, 'f1': 0.8480565371024735}
avg occupy 0.5399536083129833
39/153


eval: 100%|██████████| 216/216 [00:04<00:00, 43.48it/s]


metric score {'accuracy': 0.8382608695652174, 'f1': 0.8814279643008927}
avg occupy 0.5399536083129833
40/153


eval: 100%|██████████| 216/216 [00:05<00:00, 42.66it/s]


metric score {'accuracy': 0.8428985507246377, 'f1': 0.885605740818911}
avg occupy 0.5399536083129833
41/153


eval: 100%|██████████| 216/216 [00:05<00:00, 42.16it/s]


metric score {'accuracy': 0.8440579710144928, 'f1': 0.8865457612821593}
avg occupy 0.5399536083129833
42/153


eval: 100%|██████████| 216/216 [00:05<00:00, 41.63it/s]


metric score {'accuracy': 0.8428985507246377, 'f1': 0.8857022353437368}
avg occupy 0.5399536083129833
43/153


eval: 100%|██████████| 216/216 [00:05<00:00, 41.64it/s]


metric score {'accuracy': 0.8428985507246377, 'f1': 0.8857022353437368}
avg occupy 0.5399536083129833
44/153


eval: 100%|██████████| 216/216 [00:05<00:00, 42.46it/s]


metric score {'accuracy': 0.8428985507246377, 'f1': 0.8857022353437368}
avg occupy 0.5399536083129833
45/153


eval: 100%|██████████| 216/216 [00:05<00:00, 41.65it/s]


metric score {'accuracy': 0.8428985507246377, 'f1': 0.8857022353437368}
avg occupy 0.5399536083129833
46/153


eval: 100%|██████████| 216/216 [00:05<00:00, 42.38it/s]


metric score {'accuracy': 0.8428985507246377, 'f1': 0.8857022353437368}
avg occupy 0.5399536083129833
47/153


eval: 100%|██████████| 216/216 [00:05<00:00, 41.54it/s]


metric score {'accuracy': 0.8428985507246377, 'f1': 0.8857022353437368}
avg occupy 0.5399536083129833
48/153


eval: 100%|██████████| 216/216 [00:05<00:00, 42.20it/s]


metric score {'accuracy': 0.8428985507246377, 'f1': 0.8857022353437368}
avg occupy 0.5399536083129833
49/153


eval: 100%|██████████| 216/216 [00:05<00:00, 41.60it/s]


metric score {'accuracy': 0.8428985507246377, 'f1': 0.8857022353437368}
avg occupy 0.5399536083129833
50/153


eval: 100%|██████████| 216/216 [00:05<00:00, 42.31it/s]


metric score {'accuracy': 0.8428985507246377, 'f1': 0.8857022353437368}
avg occupy 0.5399536083129833
51/153
Trainer: qnli


WikitextBatchLoader: worker_main
WikitextBatchLoader: worker_main
WikitextBatchLoader: worker_main
WikitextBatchLoader: worker_main
Trainer: Save checkpoint path saves/glue-qnli-16-wiki-b5.pth


eval: 100%|██████████| 683/683 [00:12<00:00, 53.61it/s]


metric score {'accuracy': 0.9154310818231741}
avg occupy 0.30243502260079336


eval: 100%|██████████| 683/683 [00:17<00:00, 39.97it/s]


metric score {'accuracy': 0.5680029287937031}
avg occupy 0.30243502260079336
52/153


eval: 100%|██████████| 683/683 [00:17<00:00, 39.32it/s]


metric score {'accuracy': 0.8295808164012447}
avg occupy 0.30243502260079336
53/153


eval: 100%|██████████| 683/683 [00:17<00:00, 38.10it/s]


metric score {'accuracy': 0.8965769723595094}
avg occupy 0.30243502260079336
54/153


eval: 100%|██████████| 683/683 [00:19<00:00, 35.53it/s]


metric score {'accuracy': 0.9099395936298736}
avg occupy 0.30243502260079336
55/153


eval: 100%|██████████| 683/683 [00:19<00:00, 34.35it/s]


metric score {'accuracy': 0.9134175361522973}
avg occupy 0.30243502260079336
56/153


eval:  66%|██████▌   | 450/683 [00:14<00:07, 30.43it/s]


KeyboardInterrupt: 

In [ ]:
import pickle
import pandas as pd

with open(PICKLE_PATH, 'rb') as f:
    results = pickle.load(f)
with open('glue_benchmark.pkl', 'rb') as f:
    results_original = pickle.load(f)

def convert_data_to_results(results):
    data = []
    subsets = list(results.keys())
    factors = list(results[subsets[0]].keys())
    for factor in factors:             
        row = []
        for subset in subsets:
            row.append(results[subset][factor])
        data.append(row)
    return data, factors

data, factors = convert_data_to_results(results)
data_origin, _ = convert_data_to_results(results_original)
data_wo_data_augment = ("0.53388 (k:0.48)	0.81559 (k:0.30)	0.74667 (k:0.66)	0.90445 (k:0.41)	"+\
    "0.90893 (k:0.43)	0.70758 (k:0.38)	0.92202 (k:0.69)	0.86572 (k:0.49)	0.56338 (k:0.54)").split("\t")

factors[1] = "dynamic:w_augment"
indicies = factors+["dynamic:original", "dynamic:wo_augment"]
columns = subsets[:]
df_data = data + [data_origin[9], data_wo_data_augment]

#calculate reproducibility
data_scalar = []
for line in df_data:
    xs = []
    for item in line:
        xs.append(float(item.split()[0]))
    data_scalar.append(xs)
reproducibilities = []
for i in range(len(data_scalar)):
    rsum = 0
    for k in range(len(data_scalar[i])):
        rsum += data_scalar[i][k]/data_scalar[0][k]
    rsum /= len(data_scalar[i])
    reproducibilities.append(rsum)
for i, r in enumerate(reproducibilities):
    df_data[i].append(f"{r*100:.2f}")
columns.append("reproduce")

df = pd.DataFrame(
    df_data,
    columns=columns, 
    index=indicies
)
df = df.reindex(["bert", "dynamic:original", "dynamic:w_augment", "dynamic:wo_augment"])
tex = df.to_latex()
with open(TEX_PATH, 'w') as f:
    f.write(tex)
df

/tmp/ipykernel_159507/610143257.py:54: FutureWarning: In future versions `DataFrame.to_latex` is expected to utilise the base implementation of `Styler.to_latex` for formatting and rendering. The arguments signature may therefore change. It is recommended instead to use `DataFrame.style.to_latex` which also contains additional functionality.
  tex = df.to_latex()


cola              mnli              mrpc  \
bert                         0.53388           0.84198           0.84406   
dynamic:original    0.53388 (k:0.48)  0.84004 (k:0.33)  0.83942 (k:0.71)   
dynamic:w_augment   0.53388 (k:0.48)  0.83647 (k:0.33)  0.83826 (k:0.72)   
dynamic:wo_augment  0.53388 (k:0.48)  0.81559 (k:0.30)  0.74667 (k:0.66)   

                                qnli               qqp               rte  \
bert                         0.91543           0.90908           0.72563   
dynamic:original    0.91287 (k:0.44)  0.90920 (k:0.45)  0.72563 (k:0.55)   
dynamic:w_augment   0.91506 (k:0.48)  0.90905 (k:0.45)  0.66426 (k:0.40)   
dynamic:wo_augment  0.90445 (k:0.41)  0.90893 (k:0.43)  0.70758 (k:0.38)   

                                sst2              stsb              wnli  \
bert                         0.92431           0.88047           0.56338   
dynamic:original    0.92317 (k:0.73)  0.88043 (k:0.51)  0.56338 (k:0.63)   
dynamic:w_augment   0.92202 (k:0.68)  0.88034 (k:0.51)  0.56338 (k:0.61)   
dynamic:wo_augment  0.92202 (k:0.69)  0.86572 (k:0.49)  0.56338 (k:0.54)   

                   reproduce  
bert                  100.00  
dynamic:original       99.87  
dynamic:w_augment      98.88  
dynamic:wo_augment     97.74

# pre data augment

cola (matthews_correlation)	mnli (acc)	mrpc (acc)	qnli (acc)	qqp (acc)	rte (acc)	sst2 (acc)	stsb (pearson)	wnli (acc)

bert	0.53388	0.84198	0.84406	0.91543	0.90908	0.72563	0.92431	0.88047	0.56338

dynamic	0.53388 (k:0.48)	0.81559 (k:0.30)	0.74667 (k:0.66)	0.90445 (k:0.41)	0.90893 (k:0.43)	0.70758 (k:0.38)	0.92202 (k:0.69)	0.86572 (k:0.49)	0.56338 (k:0.54)

dynamic:original	0.53388 (k:0.48)	0.84004 (k:0.33)	0.83942 (k:0.71)	0.91287 (k:0.44)	0.90920 (k:0.45)	0.72563 (k:0.55)	0.92317 (k:0.73)	0.88043 (k:0.51)	0.56338 (k:0.63)